plumbing/protocol/packp/srvresp.go

```diff
@@ -135,17 +122,68 @@ func (r *ServerResponse) decodeACKLine(line []byte) error {
 }
 
 // Encode encodes the ServerResponse into a writer.
-func (r *ServerResponse) Encode(w io.Writer, isMultiACK bool) error {
-	if len(r.ACKs) > 1 && !isMultiACK {
-		// For further information, refer to comments in the Decode func above.
-		return errors.New("multi_ack and multi_ack_detailed are not supported")
+func (r *ServerResponse) Encode(w io.Writer) error {
+	multiAck := r.req.Capabilities.Supports(capability.MultiACK)
+	singleAckSent := false
+	commonHash := plumbing.ZeroHash
+	for cmd := range r.req.UploadPackCommands {
+		if cmd.Done {
+			if commonHash.IsZero() {
+				for _, h := range cmd.Acks {
+					if h.IsCommon && commonHash.IsZero() {
+						commonHash = h.Hash
+					}
+				}
+			}
+			continue
+		}
+		if len(cmd.Acks) == 0 {
+			os.Stdout.WriteString("!! Send nak\n")
```

Comment:
We will need to remove the debugging code using `os.Stdout` before we merge this.
Out of curiosity, did you know you can trace the pack line by setting `GIT_TRACE_PACKET=true`?
Like so:

`GIT_TRACE_PACKET=true go run _examples/memory/main.go  https://github.com/go-git/go-git`

There was a slight issue that the binary data was also being displayed, but that should soon be fixed on the latest version in the `v6-exp`.

Reply:
Thanks for the tip, I used it extensively :100: